zulipterminal/model.py

```diff
@@ -473,6 +473,36 @@ def append_message(self, event: Event) -> None:
                 last_message = self.msg_list.log[-1].original_widget.message
             else:
                 last_message = None
+            # We replace the dummy message in the current narrow
+            # if the incoming message is meant for the same narrow.
+            dummy_message_found = False
+            if self.msg_list.log:
+                possible_dummy_message = \
+                    self.msg_list.log[0].original_widget.message
+                if possible_dummy_message.get('id') is None:
+                    dummy_message_found = True
+                    if possible_dummy_message.get('type') == 'private' and \
+                            response.get('type') == 'private':
+                        if (possible_dummy_message.get('display_recipient')[0]
+                            ['email'] in [recipient['email'] for recipient
+                                          in response['display_recipient']]):
+                            recipient_count = len(
+                                response['display_recipient'])
+                            # Avoid group PM's replacing dummy message in
+                            # user PM narrow. (recipient_count > 2).
+                            # Replace dummy message when self-messaging
+                            # for the first time (recipient_count = 1).
+                            if recipient_count == 2 or recipient_count == 1:
+                                dummy_message_found = False
+                                del self.msg_list.log[0]
+
+                    if possible_dummy_message.get('type') == 'stream' and \
+                            response.get('type') == 'stream':
+                        if possible_dummy_message.get('display_recipient') == \
+                                response['display_recipient']:
+                            dummy_message_found = False
+                            del self.msg_list.log[0]
```

Comment:
This is probably the code that feels the most like it isn't quite there yet. The tests don't cover multiple recipients, which might make this clearer and cover more cases? Also, if you could integrate it with the cases below, which should cover existing cases, then that may make it easier? I'm not sure if you tried that first?

Reply:
Yup. I have integrated it with the cases below. Further, few other PRs submitted (#362, #364 and #365) have helped simplify the complexity of this commit to a great degree.